### DISCUSSION TITLE
How to query json field  using integer array in sequelize?

### DISCUSSION THREAD

**Discussion by PawanKr3012** [sequelize/sequelize#16373] [Help & Questions]
I have a model called ``Gift``. It has a field ``categories``.
Here is the model definition for ``categories`` field

```
      categories: {
            allowNull: true,
            type: DataTypes.JSON,
       },

```
Here is how categories field being populated in the database

![Screenshot from 2023-08-11 12-27-50](https://github.com/sequelize/sequelize/assets/128390744/6b99f2b7-958c-49f2-9e7d-1f974e6f3972)

Lets say i want to get entries from Gifts table which contains category 13 and 9.
From request i am getting an array of category ids.

I tried to achieve the above task using the below code.

```
const gift= await Gift.findAll({
        where:{categories: {[Op.contains]: cast(req.query.category_id, "json") }}
})

```
But i get error as ``cannot cast type text[] to json``. 

Please guide me on how to query a ``json`` field using ``array`` input.

**Comment by ephys** [comment]
Which database are you using?

**Reply by PawanKr3012** [reply]
I am using postgresql as the database

**Reply by ephys** [reply]
You can use the `to_jsonb` SQL function to convert an array to JSON

The contains operator is only available for JSONB columns, not JSON, you will also need to cast the column to JSONB (or change its type to JSONB)

```ts
// Example for Sequelize 6. Not tested and where is buggy in Sequelize 6, may or may not work:

await Gift.findAll({
  where: where(
    cast(col('categories'), 'JSONB'), 
    Op.contains, 
    fn('to_jsonb', req.query.category_id),
  ),
});

// Example for Sequelize 7:

await Gift.findAll({
  where: sql`${sql.attribute('categories')}::jsonb @> to_jsonb(${req.query.category_id})`,
});
```

If you change the type of your column to JSONB instead of casting, it's a lot easier:

```ts
await Gift.findAll({
  where: {
    categories: { [Op.contains]: sql`to_jsonb(${req.query.category_id})` }
  },
});
```

You could also consider using an integer array instead of JSON for this column, as it seems that's all you're storing in your column, and it will make your query simpler and faster. In which case the query would be simply

```ts
await Gift.findAll({
  where: {
    categories: { [Op.contains]: req.query.category_id }
  },
});
```
